### PR TITLE
Add Stateful ip field to instance group manager

### PR DESCRIPTION
**Pull request by hao-nan-li**
Add internal IP and external IP. 
Fix issues based on self service work https://github.com/GoogleCloudPlatform/magic-modules/compare/main...askubis:magic-modules:AddSupportToStatefulIPs# 

Resolved https://github.com/hashicorp/terraform-provider-google/issues/10908

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
Add Stateful IP fields to instance group manager

```

### CODE REVIEW

**Comment by hao-nan-li** [comment]
I do see some indentation problems, which I will resolve if there's other updates to do.
Create/Delete tests are in place.

**Comment by hao-nan-li** [comment]
> @hao-nan-li I see this is a WIP - is it ready for review?

Good catch! I forgot to remove [WIP].

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1037]: I see this maintains the old behavior (always send disks if len > 0) and I'm a little hesitant about suggesting changing that - but also, if the preservedState.NullFields approach is preferred for stateful ips, maybe we should actually be doing that with stateful disks as well? What do you think?

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1037]: Or alternately we could make the new fields match this behavior? Basically it seems like they should probably be consistent?

File: mmv1/third_party/terraform/resources/resource_compute_region_instance_group_manager.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_compute_region_instance_group_manager.go.erb:753]: This will cause expandStatefulPolicy to be run twice in beta currently - should be merged with the other stateful_disk block in this function.

**Review by melinath — CHANGES_REQUESTED** [review]
Requested some changes / asked some questions below. Additionally, please:

- add tests for region_instance_group_manager
- add the new fields to region_instance_group_manager
- update the release notes to match the release notes guide (with a separate code block for each resource changed)

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:322]: I need to check for this.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: I thought by setting computed and optional to true, we are basically doing the equivalent of default_from_api. I could be wrong as I was referring to https://github.com/GoogleCloudPlatform/magic-modules/blob/5268d03a79ba5644c9b8252175f642245881c8b7/mmv1/overrides/terraform/property_override.rb#L59

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:347]: Same comment as the above.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:354]: Sure I will remove the default value.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:371]: Here's where I also had concerns. It looks like a null value was expected during terraform plan phase for stateful_disk. I'm going to remove this for now and focus on finalizing changes for the other comments.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1037]: I'm a little hesitant in making changes to existing logic. I'll take a deeper look and see if I can come up with something.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: you are absolutely correct, my bad. That being said, does that API have a default that it sets?

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: The expected value is null for both internal and external ips.

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: if the field is not set in TF, and the api returns null, that sounds like a reasonable return value that shouldn't cause a permadiff - does it? Can we get around it with a customize diff if so?

Basically, if we can avoid making it O+C that's better because we'll be able to accurately report what the end state will be to users (instead of saying `(known after apply)`)

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: I can take a look into a customize diff.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:829]: Will do for this and for compute_region

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: To add notes here: If I change typeSet to typeList, when running tests for region_instance, a failure will occur saying it was expecting a Set. Swapping back from List to Set will solve the error.

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: Why is it expected to be a Set?

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: The latest log I have shows:
**2022/11/15 12:47:17 [DEBUG] setting computed for "distribution_policy_zones" from ComputedKeys
2022/11/15 12:47:17 [DEBUG] setting computed for "status" from ComputedKeys
panic: interface conversion: interface {} is []interface {}, not *schema.Set**

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: This looks like a internal type error - since this is a handwritten file, there's a good chance there's something in this file that expects this to be a Set rather than a list. It's probably not going to be solvable with API docs, unfortunately - the API wouldn't know what a schema.Set is, since that's a Hashicorp type.

Does the panic include a stack trace? That might help diagnose where the panic is coming from, which would help in evaluating whether the code can / should be changed.

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: I think I got this resolved, now fixing some importStateVerify errors

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: Looks like a customize diff is not ideal according to Tf external chat, I will leave it as it is.

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: just to have it recorded here - customize diff / diff suppress should be fine for this case since it would only be handling the empty return value; this is not a field that is partially managed by the API.

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: For Region instance manager, the log: https://paste.googleplex.com/4890762578755584, looks like the current error I'm fighting against. I'll try the other one

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: ok after fixing the error for region instance manager, it looks like I'm still getting 
stateful_external_ip {
  |           - delete_rule    = "NEVER" -> null
  |           - interface_name = "nic0" -> null
  |         }

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: interesting - I see that in paste/5070583447420928 the POST that first creates the instance group already includes values in the `externalIPs` and `internalIPs` fields.

That implies that they're in the Terraform configuration, since they're being sent on Create - but then the test fails with what looks like: a permadiff from those fields being set by the API but _not_ in the terraform configuration. Why is that happening?

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: Good point. This error also happens before the DELETE operation. 
Can we know whether the IPs are stored in the TF state file?

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: it's not easy to do from the tests; you might need to set up a local directory where you can set up files and apply them and then examine the results.

See https://googlecloudplatform.github.io/magic-modules/#using-released-terraform-binary-with-local-provider-binary for how to build & use the provider binary locally.

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:324]: Sure I can take a try.

**Review by melinath — CHANGES_REQUESTED** [review]
looks like there are some merge conflicts - not sure if this is otherwise ready for review?

**Comment by hao-nan-li** [comment]
> looks like there are some merge conflicts - not sure if this is otherwise ready for review?

I'm still working on the customized diff as we've discussed in the comments. For all conflicts, we should take what's from this commit. Will mark this as in draft until it's ready.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: Pushed a draft commit with fixed tests and additional log.print(). 
There is an additional null patching step that caused some problem. 

From what I researched, a customize diff is already implemented in place. Perhaps some details are not right at the moment.

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: Analyzed deeper into this, the issue is that during the step 5 of the test, a patch request with
"statefulPolicy": {
  "preservedState": {}
 }
is sent to the API. However the stateful_ips and the stateful_disk are not removed, therefore the API returns a value whereas TF plan sends a null value.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: You're right, I removed this field.

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb]: Sounds good, let's leave it as it is now.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1061]: I tried with NullFields approach and didn't notice a difference. I prefer leaving it as it is for now.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1047]: as discussed - we should remove this expander and just use `d.Get` - simpler. If there's a need to suppress a diff we should use a set hash func or diff suppress func rather than trying to handle it here.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:847]: it would great to simplify these - for example:
```
<% unless version == "ga" -%>
	if d.HasChange("stateful_disk") || d.HasChange("stateful_internal_ips") || d.HasChange("stateful_external_ips") {
<% else -%>
        if d.HasChange("stateful_disk") {
<% end -%>
```

That would let us just have one block for expanding this field.

**Review by melinath — CHANGES_REQUESTED** [review]
Overall - let's try to simplify this code and then go through & troubleshoot it. Once you've done that, here are a couple scenarios to test & collect sample request/responses for:

- Creating a new instance group manager with stateful ips vs without
- Updating an IGM to remove stateful IPs & disks at the same time
- Updating an IGM to remove stateful IPs but not disks
- Updating an IGM to remove disks but not stateful IPs

It would also be good to double-check the API SDK docs to figure out what NullFields is supposed to do just in case.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1045]: I did the changes and did some ForceSendFields to delete the fields. As far as now, the disk could be deleted, but not the stateful Ips. https://paste.googleplex.com/6460650149642240 

According to the log above, the patch request is sent, the disk is deleted, not the Ips. My suggestion is to add the Ips in the verifyStateIgnore. And not mark these fields as Computed

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:1045]: Just to have it written here as well - this seems like an API bug because:

- Sending empty objects deletes disks but not ips - we'd expect it to be consistent
- The end state of the API object is the same as what was set by the test - it doesn't seem to be getting reset to a default state by the API (so it's probably not an optional + computed field)

Since it's a bug in the API we can't just ignore the read value - that would mean that users think a change has gone through but actually the value on the API side is different from what's in Terraform.

**Comment by hao-nan-li** [comment]
https://yaqs.corp.google.com/eng/q/3181188405341978624 referring to the error.

If the error above takes long time to resolve, could we merge the PR without a delete test? As of now create/update works

**Comment by melinath** [comment]
We can't merge this if the fields can't be updated properly; otherwise this will surface as a bug.

**Review by rileykarson — COMMENTED** [review]
Based on talking with @hao-nan-li and a brief investigation on my end, the implementation of PATCH in this API is nonstandard, and doesn't follow common conventions or the conventions shared at aip.dev. That's surfaced a limitation in the client libraries that's blocking us- so let's ship this without update support for the fields at all.

There's already a workaround- users will be able to manage the values out-of-band to update it, and then just match up their Terraform configuration. To avoid this situation in the future, the API should follow standard GCP conventions.

**Comment by rileykarson** [comment]
https://github.com/googleapis/google-api-go-client/releases/tag/v0.105.0 includes a fix for the update problem. I'd suggest deferring update support to a followup PR, rather than working it in to this one, since we'll need to do a client library update to make use of it.

File: mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb:371]:
```suggestion
	network := fmt.Sprintf("tf-test-igm-%s", randString(t, 10))
```

File: mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb:387]:
```suggestion
			{
```

File: mmv1/third_party/terraform/tests/resource_compute_region_instance_group_manager_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_compute_region_instance_group_manager_test.go.erb:377]:
```suggestion
	network := fmt.Sprintf("tf-test-igm-%s", randString(t, 10))
```

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: @rileykarson Just want to double-check that when you say "no update support" you mean "we ship it as-is & if they want to remove nics they do it manually" rather than "we make these new fields ForceNew for now and change that to allow in-place updates later."

File: mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb
Comment (melinath) [inline mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb:1523]: It looks like `interface_name` is updatable on both internal and external ips - but changing that would probably be seen as a "deletion" by the API because its data structure is different. So, just noting this for myself / future folks - nothing to change here.

**Review by melinath — APPROVED** [review]
lgtm. Noted a few indentation cleanups below but nothing blocking unless this should be using forcenew - but it sounds from Riley's comment like we're just accepting there's a bug in the update logic on the API side for now. 👍

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (rileykarson) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: No update support means the second: "we make these new fields ForceNew for now and change that to allow in-place updates later."

Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: So based on the current PR, we add **ForceNew: true** to the statefulIps, and in another PR, we allow in-place updates.

Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: @hao-nan-li yep.

**Review by melinath — CHANGES_REQUESTED** [review]
see above discussion

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: Just want to make sure, it's an update test required if we set the field to forceNew?

File: mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb
Comment (hao-nan-li) [reply mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb:371]: Thanks for pointing them out, locally I really don't see the indentation difference. Will try to correct them properly.

File: mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb
Comment (melinath) [reply mmv1/third_party/terraform/resources/resource_compute_instance_group_manager.go.erb:329]: no, that's not required. in fact the fields need to actively _not_ be changed in an update test - otherwise the test will pass but not exercise the update-in-place path.

If the resource doesn't already have it, you could consider using lifecycle.prevent_destroy to ensure that the test doesn't accidentally delete the resource: https://github.com/hashicorp/terraform-provider-google/issues/12594

**Review by melinath — APPROVED** [review]
LGTM - there are a few indentation nits listed above but they're not blockers.

**Review by melinath — APPROVED** [review]

**Comment by shuyama1** [comment]
Looks like we still need to add documentation for this feature manually since this is a handwritten resource.

**Comment by hao-nan-li** [comment]
> Looks like we still need to add documentation for this feature manually, since this is a handwritten resource.

Sure I can add it with https://github.com/GoogleCloudPlatform/magic-modules/pull/7075

**Pull request merged**
